lib/clerk/authenticatable.rb

```diff
@@ -5,16 +5,31 @@ module Authenticatable
     extend ActiveSupport::Concern
 
     protected
-    def clerk_session
-      request.env["clerk"].session
+
+    # Involves a network call
+    def clerk_verify_session
+      request.env["clerk"].verify_session
+    end
+
+    def clerk_session_claims
+      request.env["clerk"].session_claims
     end
 
+    def clerk_session_token
+      request.env["clerk"].session_token
+    end
+
+    # Involves a network call
     def clerk_user
       request.env["clerk"].user
     end
 
+    def clerk_user_id
+      request.env["clerk"].user_id
+    end
+
     def clerk_user_signed_in?
-      !!clerk_user
+      !!clerk_session_claims
     end
 
```

Comment:
The summary of this API is:

- `clerk_verify_session` verifies the session denoted by the session JWT. It always involves a network roundtrip if the user is logged-in, since it will always query BAPI
- `clerk_session_claims` returns the decoded claims of the *verified* session token
- `clerk_session_token` returns the *verified* token in raw format
- `clerk_user` returns the user object of the *verified* token. It queries BAPI once for each provided token (subsequent calls will return the cached object)
- `clerk_user_signed_in?` doesn't hit the network, ever

How do you all feel with this API? @SokratisVidros @colinclerk @chanioxaris

Reply:
It sounds like clerk_user.id will trigger a a network request

can we offer an intuitive way to retrieve the user id that won’t trigger a network request?  I don’t think most devs will necessarily be aware of the token implementation detail

clerk_user_id should work

otherwise, we can lazy load clerk_user and clerk_session attributes, so fields that don’t require a network request are prepopulated

Reply:
Oh there's already `clerk_user_id`, which I forgot to mention.

Reply:
Awesome

Reply:
Something about this is still confusing me but I don't have a path forward right now.

The existence of `clerk_verify_session` is implying that `clerk_session_claims` are somehow unverified.  I don't think is intentional, just a result of the naming.

I believe `clerk_verify_session` provides network-bound extra security, which will confirm that the session is still active even if the token hasn't expired.  In normal operation this will never change the outcome of authentication, but if there's a malicious attacker it might (for a maximum of 1 minute).

Reply:
Yes, it's confusing, I agree.

How about:

- `clerk_verified_session_claims`
- `clerk_verified_session_token`
- `clerk_force_verify_session` (or `clerk_force_verify_session!`)

And also to remedy this confusion, all these methods should be documented and state explicitly that they hit the network (for `force_verify`)

Reply:
Would `reverified` be good for indicating the double verification?

```
clerk_session_token = some kind of token object?
clerk_session_token.raw = unverified string
clerk_session_token.claims = verified session token claims
clerk_session_token.unsafe_claims = unverified session token claims
clerk_user_id = clerk_session_token_claims.sub
clerk_session_id = clerk_session_token_claims.sid


clerk_reverify_session!  # In Ruby, this can be a before action. In Node, this can be `withReverifiedSession` or `requireReverifiedSession`

clerk_session = network-retrieved Session object
clerk_user = network-retrieved Session object
```

Reply:
Yes, I like `#clerk_reverify_session!`. It's a normal helper that can be used in controllers (so in before actions) or in views. Do you think we need anything more than that?

I'd propose the following API then (all the following are helpers):

- `clerk_verified_session_claims`
- `clerk_verified_session_token`
- `clerk_reverify_session!`
- `clerk_user`
- `clerk_user_id`

WDYT?